arch/x86/compare258_avx.c

```diff
@@ -0,0 +1,60 @@
+/* compare258_avx.c -- AVX2 version of compare258
+ * Copyright Mika T. Lindqvist  <postmaster@raasu.org>
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#include "../../zbuild.h"
+#include "../../zutil.h"
+
+#include "fallback_builtins.h"
+
+#if defined(X86_AVX2) && defined(HAVE_BUILTIN_CTZ)
+
+#include <immintrin.h>
+#ifdef _MSC_VER
+#  include <nmmintrin.h>
+#endif
+
+// UNALIGNED_OK, AVX2 instrinic comparison
+static inline int32_t compare258_unaligned_avx_static(const unsigned char *src0, const unsigned char *src1) {
+    register const unsigned char *src0start = src0;
+    register const unsigned char *src0end = src0 + 258;
+    __m256i ymm_zero = _mm256_set1_epi8(0);
+
+    if (*(uint16_t *)src0 != *(uint16_t *)src1)
+        return (*src0 == *src1);
+
+    src0 += 2, src1 += 2;
+    if (*src0 != *src1)
+        return 2;
+    if (src0[1] != src1[1])
+        return 3;
+
+    do {
+        __m256i ymm_src0, ymm_src1, ymm_xor;
+        ymm_src0 = _mm256_loadu_si256((__m256i*)src0);
+        ymm_src1 = _mm256_loadu_si256((__m256i*)src1);
+        ymm_xor = _mm256_xor_si256(ymm_src0, ymm_src1); // identical bytes = 00
+        ymm_xor = _mm256_cmpeq_epi8(ymm_xor, ymm_zero); // non-identical bytes = 00, identical bytes = FF
+        int mask = _mm256_movemask_epi8(ymm_xor); 
+        if (mask != 0xFFFFFFFF) {
```

Comment:
/arch/x86/compare258_avx.c:40:18: warning: comparison of integer expressions of different signedness: ‘int’ and ‘unsigned int’ [-Wsign-compare]

Reply:
0xFFFFFFFF is same as INT32_C(-1) but in hexadecimal it is more obvious why it needs to be that specific number.